created make.bat in doc folder

me-t1me opened the pull request:
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
e.g. [#22872(comment)](https://github.com/sympy/sympy/pull/22872#issuecomment-1030665199)
e.g. [#17817(comment)](https://github.com/sympy/sympy/issues/17817#issuecomment-548507161)
fixes #16460
fixes #17817
#### Brief description of what is fixed or changed
Created make.bat file in doc folder, so that `make html` works in windows

#### Other comments
To run `make html`, in cmd administrator or powershell run `choco install make`
chocolatey is required for above command
#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

me-t1me commented:
@oscarbenjamin ping for review

me-t1me commented:
Should I add instruction as comment in `make.bat`file?

me-t1me commented:
I have no idea who should I ping for review

oscarbenjamin commented:
> I have no idea who should I ping for review

It isn't always necessary to ping people.

I'm sure someone who uses Windows and is interested in ensuring that it is easy to build the docs on Windows will see this and can try this out.

> Should I add instruction as comment in `make.bat`file?

I was thinking more like here:
https://docs.sympy.org/latest/documentation-style-guide.html#installation

me-t1me commented:
chocolatey is a command-line package manager, I will create instructions for installing chocolatey then installing make and other packages using `choco install`

oscarbenjamin commented:
Why is chocolatey needed?

The `sphinx_math_dollar` extension is needed and can be installed as `pip install sphinx-math-dollar`.

me-t1me commented:
make command is not an internal command in command prompt so it is installed using chocolatey

smichr commented:
I pip installed each thing that was reported as missing
```
...\sympy\doc>make html
Running Sphinx v4.4.0
loading translations [en]... done

Extension error:
Could not import extension sphinx_math_dollar (exception: No module named 'sphin
x_math_dollar')

...\sympy\doc>make html
Running Sphinx v4.4.0
loading translations [en]... done

Extension error:
Could not import extension sphinx_reredirects (exception: No module named 'sphin
x_reredirects')

...\sympy\doc>make html
Running Sphinx v4.4.0
loading translations [en]... done

Extension error:
Could not import extension myst_parser (exception: No module named 'myst_parser'
)

...\sympy\doc>make html
Running Sphinx v4.4.0
loading translations [en]... done
making output directory... done
WARNING: favicon file '../_build/logo/sympy-notailtext-favicon.ico' does not exi
st
myst v0.16.1: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_ex
tensions=['dollarmath', 'linkify'], linkify_fuzzy_links=True, dmath_allow_labels
=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=Fals
e, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|ou
tput_area', disable_syntax=[], url_schemes=['http', 'https', 'mailto', 'ftp'], h
eading_anchors=2, heading_slug_func=None, html_meta=[], footnote_transition=True
, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 270 source files that are out of date
updating environment: [new config] 270 added, 0 changed, 0 removed
Creating redirect file 'install.html' pointing to '{new_target}'.
Creating redirect file 'documentation-style-guide.html' pointing to '{new_target
}'.
Creating redirect file 'gotchas.html' pointing to '{new_target}'.
Creating redirect file 'special_topics\classification.html' pointing to '{new_ta
rget}'.
Creating redirect file 'special_topics\finite_diff_derivatives.html' pointing to
 '{new_target}'.
Creating redirect file 'special_topics\intro.html' pointing to '{new_target}'.
Creating redirect file 'special_topics\index.html' pointing to '{new_target}'.
Creating redirect file 'modules\index.html' pointing to '{new_target}'.
Creating redirect file 'modules\physics\index.html' pointing to '{new_target}'.

Exception occurred:
  File "c:\python37-32\lib\site-packages\markdown_it\rules_core\linkify.py", lin
e 31, in linkify
    raise ModuleNotFoundError("Linkify enabled but not installed.")
ModuleNotFoundError: Linkify enabled but not installed.
The full traceback has been saved in ...\AppData\Local\Temp\sphinx-e
rr-ab0fhv6_.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message c
an be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphin
x/issues>. Thanks!
```
I followed directions for Powershell installation of choco*
```
...\sympy\doc>powershell
Windows PowerShell
Copyright (C) 2009 Microsoft Corporation. All rights reserved.

PS ...\sympy\doc> Get-ExecutionPolicy
Restricted
PS ...\sympy\doc> Set-ExecutionPolicy AllSigned
Set-ExecutionPolicy : Access to the registry key 'HKEY_LOCAL_MACHINE\SOFTWARE\M
icrosoft\PowerShell\1\ShellIds\Microsoft.PowerShell' is denied.
At line:1 char:20
+ Set-ExecutionPolicy <<<<  AllSigned
    + CategoryInfo          : NotSpecified: (:) [Set-ExecutionPolicy], Unautho
   rizedAccessException
    + FullyQualifiedErrorId : System.UnauthorizedAccessException,Microsoft.Pow
   erShell.Commands.SetExecutionPolicyCommand
```

eagleoflqj commented:
IMHO a change in CI config is needed, to include a Windows Server host for building docs.
I use chocolatey a lot, but the official winget may also fit.

oscarbenjamin commented:
> IMHO a change in CI config is needed, to include a Windows Server host for building docs.

What exactly do you mean? The docs are built in CI but this is about building them locally.

eagleoflqj commented:
> What exactly do you mean? The docs are built in CI but this is about building them locally.

Currently we only use Linux in CI. I mean setting up a Windows Server 2016 in CI, to imitate building docs locally on Windows 10.

oscarbenjamin commented:
What do you mean by "imitate"?

It would be easy to have CI output the built docs for every PR but we would very quickly run out of storage space in GA:
https://github.com/sympy/sympy/pull/22861#issuecomment-1013657785

eagleoflqj commented:
"Imitate" is just because Windows Server != Windows 10. GA only provides Windows Server so it's the only way to test anything targeting on Windows.
Since this PR only aims for "building docs successfully on Windows", it doesn't need to create any artifacts. Simply run build scripts on Windows and see if there's any error log.

oscarbenjamin commented:
> it's the only way to test anything targeting on Windows.

Oh, I see. You just mean to test if building the docs on Windows works.

I'm not sure we need to test that in CI because it's not something that users need and contributors can just report if there's a problem with it. I doubt that being able to build the docs on Windows is something that is particularly affected by changes that are routinely made to the docs.

Right now it just seems like it isn't easy to do it on Windows and what we need is a clear prescription that works. It's a long time since I used Windows but when I did I always had GNU Make installed from e.g. MSYS2/MINGW. Maybe that's not such an obvious set up these days. We should have some clear instructions that work though because many contributors use Windows.

me-t1me commented:
@smichr for choco, you have run powershell as administrator
for linkify problem `pip install linkify-it-py`
I will also add installation instruction for python dependencies in docs

me-t1me commented:
>Right now it just seems like it isn't easy to do it on Windows and what we need is a clear prescription that works. It's a long time since I used Windows but when I did I always had GNU Make installed from e.g. MSYS2/MINGW. Maybe that's not such an obvious set up these days. We should have some clear instructions that work though because many contributors use Windows.

I experience some problems while trying build-docs will MSYS2/MinGW
1. managing an environment like MinGW is always a hassle to me and a lot of errors get thrown at me for builds like these
2. file `doc/generate_logos`  will throw warnings as favicon files cannot be created by it as code written in it is originally for Linux based os systems
3. Docker also gave me a similar experience and I find copy-pasting commands easy as compared to maintaining these environments

me-t1me commented:
>Right now it just seems like it isn't easy to do it on Windows and what we need is a clear prescription that works. It's a long time since I used Windows but when I did I always had GNU Make installed from e.g. MSYS2/MINGW. Maybe that's not such an obvious set up these days. We should have some clear instructions that work though because many contributors use Windows.

I experience some problems while trying build-docs with MSYS2/MinGW
1. managing an environment like MinGW is always a hassle to me and a lot of errors get thrown at me for builds like these
2. file `doc/generate_logos`  will throw warnings as favicon files cannot be created by it as code written in it is originally for Linux based os systems
3. Docker also gave me a similar experience and I find copy-pasting commands easy as compared to maintaining these environments

me-t1me commented:
>Right now it just seems like it isn't easy to do it on Windows and what we need is a clear prescription that works. It's a long time since I used Windows but when I did I always had GNU Make installed from e.g. MSYS2/MINGW. Maybe that's not such an obvious set up these days. We should have some clear instructions that work though because many contributors use Windows.

I experience some problems while trying build-docs with MSYS2/MinGW
1. managing an environment like MinGW is always a hassle to me and a lot of errors get thrown at me for builds like these
2. file `doc/generate_logos`  will throw warnings as favicon files cannot be created by it as code is written in it is originally for Linux based os systems
3. Docker also gave me a similar experience and I find copy-pasting commands easy as compared to maintaining these environments

bjodah commented:
In my humble opinion, WSL2 is so good that I have stopped spending time trying to get things to work "natively" whenever I'm prescribed a Windows machine.

asmeurer commented on doc/make.bat:
```diff
@@ -0,0 +1,49 @@
+:html
+mkdir %BUILDDIR%\logo
+python ./generate_logos_win32.py -d
+%SPHINXBUILD% -M %1 %SOURCEDIR% %BUILDDIR% %SPHINXOPTS% %O%
```
We don't need to copy everything from the Makefile here, but we should at least include the `SPHINXOPTS=-W --keep-going` as otherwise it won't make the build fail on warnings like it should.

asmeurer replied:
No, it makes it so that build *does* fail when there are warnings (the `--keep-going` makes it show all warnings instead of just the first one).

asmeurer commented:
Why is there a separate script for generating the logos in Windows? What is the difference from generate_logo.py. I think we should just make the one script portable across different platforms.

asmeurer commented:
If this is useful to people then it's OK to have it as far as I'm concerned. I removed the make.bat because it didn't actually match the various options in the Makefile, and as far as I know no one was actually using it. But if this makes them actually match up, and you have tested it and it works, then that's fine. I agree with Oscar that we don't need to go to the trouble of setting up a Windows CI just to test this.

asmeurer commented on doc/make.bat:
```diff
@@ -0,0 +1,49 @@
+@ECHO OFF
```
Is this file based off the make.bat that sphinx-quickstart generates or is it something you cooked up yourself?

me-t1me replied:
I started with make.bat made by sphinx-quickstart, then added parameters like livehtml and clean

me-t1me commented:
@asmeurer
> Why is there a separate script for generating the logos in Windows? What is the difference from generate_logo.py. I think we should just make the one script portable across different platforms.

there is a issue in `rsvg-convert` of windows that it does not accept absolute path, so I have to change current working directory many times. I can create different functions in `doc/generate_logos` and delete `doc/generate_logos_win32`

oscarbenjamin commented:
Maybe we should just commit the converted logos to git so they don't need to be rebuilt.

me-t1me commented:
> If this is useful to people then it's OK to have it as far as I'm concerned. I removed the make.bat because it didn't actually match the various options in the Makefile, and as far as I know no one was actually using it. But if this makes them actually match up, and you have tested it and it works, then that's fine. I agree with Oscar that we don't need to go to the trouble of setting up a Windows CI just to test this.

current version of `make.bat` support these options
```cmd
Sphinx v4.4.0
Please use `make target' where target is one of
  html        to make standalone HTML files
  dirhtml     to make HTML files named index.html in directories
  singlehtml  to make a single large HTML file
  pickle      to make pickle files
  json        to make JSON files
  htmlhelp    to make HTML files and an HTML help project
  qthelp      to make HTML files and a qthelp project
  devhelp     to make HTML files and a Devhelp project
  epub        to make an epub
  latex       to make LaTeX files, you can set PAPER=a4 or PAPER=letter
  text        to make text files
  man         to make manual pages
  texinfo     to make Texinfo files
  gettext     to make PO message catalogs
  changes     to make an overview of all changed/added/deprecated items
  xml         to make Docutils-native XML files
  pseudoxml   to make pseudoxml-XML files for display purposes
  linkcheck   to check all external links for integrity
  doctest     to run all doctests embedded in the documentation (if enabled)
  coverage    to run coverage check of the documentation (if enabled)
  clean       to remove everything in the build directory
  livehost    to create and host HTML on http://localhost:3000
```

smichr commented:
> @smichr for choco, you have run powershell as administrator

running as Administrator I get
```
PS ...\sympy> Set-ExecutionPolicy Bypass -Scope Process -Force; [Sys
tem.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager
]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadSt
ring('https://community.chocolatey.org/install.ps1'))
Exception setting "SecurityProtocol": "Cannot convert value "3312" to type "Sys
tem.Net.SecurityProtocolType" due to invalid enumeration values. Specify one of
 the following enumeration values and try again. The possible enumeration value
s are "Ssl3, Tls"."
At line:1 char:85
+ Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointMan
ager]:: <<<< SecurityProtocol = [System.Net.ServicePointManager]::SecurityProto
col -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://c
ommunity.chocolatey.org/install.ps1'))
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyAssignmentException

Exception calling "DownloadString" with "1" argument(s): "The underlying connec
tion was closed: An unexpected error occurred on a send."
At line:1 char:219
+ Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointMan
ager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -b
or 3072; iex ((New-Object System.Net.WebClient).DownloadString <<<< ('https://c
ommunity.chocolatey.org/install.ps1'))
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException
```

me-t1me commented:
Maybe this issue is due to older version of powershell you can download v3 from [here](https://www.microsoft.com/en-us/download/details.aspx?id=34595)

asmeurer commented:
> current version of make.bat support these options

The only other one we support is latex, but that will probably require more manual work to support as we have some custom fixes to make latex work in the Makefile.

me-t1me commented:
Should I delete `doc/generate_logos_win32` and put all code inside `doc/generate_logos` or do what @oscarbenjamin said?

asmeurer commented:
Let's make a separate issue about commiting the logos to the repo. Either way, we should keep the script to regenerate the logos regardless of whether we commit them to the repo.

me-t1me commented:
`make latex` gives warning `a suitable image for latex builder not found: ['image/svg+xml'] (modules/vector/coordsys_rot.*)` and I not sure what is causing this problem

asmeurer commented:
The reason we don't use rsvg or any other tools is that they all convert the svgs incorrectly. Headless Chrome is the only tool that does it correctly for all the SVGs. See https://github.com/sympy/sympy/pull/22468.

me-t1me commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,84 @@
+        """Converts the image from SVG to PDF using chrome."""
+
+        HTML = "<html ><head ><style >body {margin: 0; }</style ><script >function init() {const element = document.getElementById('targetsvg');const positionInfo = element.getBoundingClientRect();const height = positionInfo.height;const width = positionInfo.width;const style = document.createElement('style');style.innerHTML = \`@page {margin: 0; size: \${width}px \${height}px}\`;document.head.ap ...
+        chrome = os.path.join(os.environ['PROGRAMW6432'],
+             "Google\Chrome\Application\chrome.exe")
```
I think this extension can also work for linux, macOS, we just need to change `chrome` variable

me-t1me commented:
@asmeurer

asmeurer commented:
Can we delete the svg-to-pdf shell script?

asmeurer commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,84 @@
+
+class Converter(ImageConverter):
+    conversion_rules = [
+        ('image/svg+xml', 'application/pdf'),
```
Does this make it so this only runs when it needs a PDF (i.e., only for the latex build)?

me-t1me replied:
We just need to change conversion-rule to run it for any other image type,  further we can change its parent class ImageConverter to get our extention to run on different file types as well,  this extension will run whenever it sees conversion requirements of (image/svg+xml, application/pdf)

me-t1me commented:
> Can we delete the svg-to-pdf shell script? 

yes

me-t1me commented:
I checked this commit for windows but I am not able to check this code for Mac and Linux.
Also I have not deleted this suffice rule from Makefile
```makefile
.svg.pdf:
	./convert-svg-to-pdf.sh $< $@
```

asmeurer commented:
If the Sphinx builder does it you can delete it from the makefile. The CI will test Linux. I can test Mac.

me-t1me commented:
How I implemented this extension for windows  -
1. I created .HTML file
2. put HTML tag with src=/path/to/svg.svg  into .HTML file
3. then used `google-chrome --headless --disable-gpu --disable-software-rasterizer --print-to-pdf=/home/runner/work/sympy/sympy/doc/_build/doctrees-latex/images/coordsys_rot.pdf /home/runner/work/sympy/sympy/doc/src/modules/vector/coordsys_rot.html`
4. then deleted .HTML file

Is there a problem with this process while running this in linux?

asmeurer commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,101 @@
+        chrome = self.chrome_command()
+        html_path = _from.replace(".svg", ".html")
+        try:
+            f = open(html_path, "w+")
```
Just `with open(...)` instead of calling `flush()` and `close()`.

asmeurer commented:
That sounds fine, although I would use the `tempfile` module to write the HTML in a temporary directory (that will work on both Windows and Linux).

me-t1me commented:
`google-chrome --headless --disable-gpu --disable-software-rasterizer --print-to-pdf=/home/runner/work/sympy/sympy/doc/_build/doctrees-latex/images/vec_rep.pdf /tmp/tmpw5o4hlk1.html`
this command is being implemented, but it is not working

me-t1me commented:
@asmeurer I am not able to make `convert-svg-to-pdf.py` extension work for Linux. I have no idea what is wrong with this extension

asmeurer commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,101 @@
+            logger.info(args)
+            logger.debug('Invoking %r ...', args)
+            p = subprocess.Popen(
+                args, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
```
You need to pass a list of arguments to subprocess, like `['google-chrome', '--headless', '--disable-gpu', ...]`. 

It also is usually simpler to use `subprocess.run` instead of `Popen`.

me-t1me commented:
I think now we can merge this pr

oscargus commented:
Can you please rebase and merge the commits?

Basically do 
```
git rebase -i master
```
Change all/most commits except one to `fixup`.

And the force push as (make sure that you get it correct before doing it, since it will overwrite all your previous changes)
```
git push origin pr_make_bat --force
```

me-t1me commented:
@asmeurer Is this fine?

oscargus commented:
25 commits is still a bit much. Especially when 10 or so are called "debugging". Keep one or a few commits with informative descriptions.

oscargus commented:
Is there a reason to try to get as many commits as possible? Adding/removing single files usually doesn't warrant a commit message. It is sometimes suitable to have more than one commit per PR, if multiple things are changed and one would like to be able to track them.

me-t1me commented:
I squashed all commit into one, is this fine?

asmeurer commented:
The rendering of the figures in the PDF is broken here. Here is an example (from page 1575). 

**master**

<img width="548" alt="Screen Shot 2022-03-02 at 2 56 24 PM" src="https://user-images.githubusercontent.com/71486/156456693-8c203b5d-cb38-4c44-b714-56e9e01d0d4b.png">

**this branch**

<img width="505" alt="Screen Shot 2022-03-02 at 2 59 09 PM" src="https://user-images.githubusercontent.com/71486/156456702-70566d3b-5efe-4db6-9aaf-4fc8aa81315a.png">

me-t1me commented:
@asmeurer

me-t1me commented:
Is this good

asmeurer commented:
I won't be able to review this until next week, but if someone else wants to, please manually build the PDF and verify the diagrams look correct, as above.

me-t1me commented:
Is it still not working as intended?

asmeurer commented:
Sorry I didn't have a chance to test this again. Let me run it now. In the meantime, there is a minor merge conflict that needs to be fixed here.

asmeurer commented on doc/generate_logos.py:
```diff
@@ -15,11 +15,15 @@
+
+if system()[0:3].lower() == "win":
+    default_source_dir = os.path.join(os.path.dirname(__file__), "src\logo")
+    default_output_dir = os.path.join(os.path.dirname(__file__), "_build\logo")
```
You don't need this if statement. The whole point of `join` is to add the system specific path separator, so you can just use `os.path.join(os.path.dirname(__file__), "src", "logo")`.

asmeurer commented on doc/generate_logos.py:
```diff
@@ -185,6 +195,53 @@ def convert_to_png(fn_source, output_dir, sizes):
+        for size in sizes:
+            fn_out = "%s-%dpx.png" % (name, size)
+            # fn_out = os.path.join(output_dir, fn_out)
+            fn_out = os.path.join('..\\..\\_build\\logo', fn_out)
```
Same here. Use `os.path.pardir` and pass the individual arguments to join.

asmeurer commented on doc/generate_logos.py:
```diff
@@ -185,6 +195,53 @@ def convert_to_png(fn_source, output_dir, sizes):
                 logging.debug("command: %s" % cmd)
                 logging.debug("return code: %s" % p.returncode)
 
+def convert_to_png_win(fn_source, output_dir, sizes):
```
What's the difference between these two functions. They look very similar. Can they be combined with only logic for the different parts split between Windows and non-Windows?

asmeurer commented on doc/src/guides/contributing/build-docs.rst:
```diff
@@ -104,6 +104,34 @@ convenience than a virtual machine if you partially need a Linux environment
+
+Installing python dependencies::
+
+   pip install sphinx_math_dollar sphinx_reredirects myst_parser linkify-it-py
```
Use `pip install -r doc/requirements.txt` (the other instances on this page have been updated to have this).

asmeurer commented:
The PDF docs appear to build correctly now. I only tested this on Mac.

eagleoflqj commented on doc/generate_logos.py:
```diff
@@ -215,14 +234,19 @@ def convert_to_ico(fn_source, output_dir, sizes):
+        if system()[0:3].lower() == "win":
+            cmd = "magick {} {}".format(" ".join(pngs), fn_out)
+        else:
+            cmd = "convert {} {}".format(" ".join(pngs), fn_out)
```
This if-else is not needed as you previously defined cmd according to os.

me-t1me replied:
Sorry for the delay, will get it done ASAP

me-t1me commented:
Doctest test failed in the last commit is not due to this pr. 

`File "/home/runner/work/sympy/sympy/sympy/combinatorics/perm_groups.py", line 3697, in sympy.combinatorics.perm_groups.PermutationGroup.schreier_sims_random
Failed example:
    _verify_bsgs(S, base, strong_gens)
Expected:
    True
Got:
    False`

Error is occurring from this line [sympy/combinatorics/perm_groups.py](https://github.com/sympy/sympy/blob/832c24fec1046eaa544a4cab4c69e3af3e651759/sympy/combinatorics/perm_groups.py#L3697l)

Should I raise an issue regarding this?

eagleoflqj commented:
There’s already #23467

me-t1me committed:
created make.bat in doc folder & a new extension convert-svg-to-pdf.py
fix for breakdown of images in references

eagleoflqj commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
```
This Python 2 style comment should be removed.

eagleoflqj commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,94 @@
+if False:
+    # For type annotation
+    from typing import Any, Dict  # NOQA
+    from sphinx.application import Sphinx  # NOQA
```
The `if` should be removed. Also, if you use `from __future__ import annotations` at the 1st line, you can just use `dict` instead of `Dict`.

eagleoflqj commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,94 @@
+    ]
+
+    def is_available(self):
+        # type: () -> bool
```
Comment-style type hint should be changed to inline-style.
e.g. `def is_available(self) -> bool:`

eagleoflqj commented on doc/ext/convert-svg-to-pdf.py:
```diff
@@ -0,0 +1,94 @@
+        """Confirms if converter is available or not."""
+        return True
+
+    def chrome_command(self):
```
This function only dispatches by different OS, but it should provide a usable command like the deleted script did. And if the command isn't available, it should exit with an error meesage, instead of a warning and proceed.
Moreover, I don't think we should force users to install proprietary Google Chrome. At least `chromium` command should be tried first of all.

me-t1me replied:
I am not sure if all chromiums are able to perform this task in headless mode. @asmeurer implemented this thing in shell script and I only converted it into python script to run on any OS

> implemented this thing in shell script and I only converted it into python script to run on any OS
I changed logger.warning to logger.error

asmeurer replied:
I didn't test chromium (I only did minimal testing on Linux). If it can work, I agree we should try it too. My guess is that it will work since the SVG code shouldn't be part of the Google specific parts of Chrome.

me-t1me committed:
added chromium options

eagleoflqj commented:
The chromium issue isn't fix. I'll take care of it and test on Ubuntu, macOS and Windows 10.

eagleoflqj committed:
fix chromium
Tested on:
- Ubuntu 22.04 Chromium
- Windows 10 Chromium
- MacOS 12.4 Chromium
- Ubuntu 20.04 Chrome
- Windows 11 Chrome
- MacOS 12.2.1 Chrome

eagleoflqj commented:
> The PDF docs appear to build correctly now. I only tested this on Mac.

I'm getting `! Package fontspec Error: The font "DejaVu Serif" cannot be found.` when running `make pdf` on macOS, using SymPy master. I installed all dependencies. How to resolve that?

asmeurer commented:
You need to have the DejaVu fonts installed. I didn't realize we had that as a dependency for the PDF but it makes sense because we typically use DejaVu Sans Mono for the Unicode pretty printed output. I guess we need to update the dev docs https://docs.sympy.org/dev/guides/contributing/build-docs.html#pdf-documentation (unless it's included in the texlive-fonts-recommended package).

eagleoflqj commented:
Tested building pdf on `{Ubuntu, Windows, macOS}×{Chromium, Chrome}`.

asmeurer commented:
Did you test that the SVG diagrams are rendered correctly?

asmeurer commented:
I just tested it on my Mac and it works and the diagrams are rendered correctly. I'll take your word for Windows and Linux.

Let's open an issue for splitting doc/ext/convert-svg-to-pdf.py into a separate Sphinx extension that can be used by other projects.

eagleoflqj commented:
> Did you test that the SVG diagrams are rendered correctly?

Yes, I checked https://github.com/sympy/sympy/pull/23035#issuecomment-1057438350 for every built pdf.

> we should add them to .gitignore.

Hmm, I've already done that in this PR. I guess you are in an old branch that hasn't merged current master.

asmeurer commented:
Oh I see. It must have happened when I switched back to another branch.
